Overhauling SingleListGrader answer structure, validation and grading

jolyonb opened the pull request:
This PR updates the answer validation for SingleListGrader. It actually leverages `ItemGrader` in a neat way that makes everything nicer, in my opinion. Addresses #247 for SingleListGrader (but not ListGrader, which is a whole 'nother beast!).

The code:
```python
grader = SingleListGrader(
    answers=['a', 'b'],
    subgrader=StringGrader()
)
pprint.pprint(grader.config['answers'])
```
currently produces:
```python
([({u'expect': u'a', u'grade_decimal': 1, u'msg': u'', u'ok': True},),
  ({u'expect': u'b', u'grade_decimal': 1, u'msg': u'', u'ok': True},)],)
```
whereas in the overhaul, it produces:
```python
({u'expect': [({u'expect': u'a',
                u'grade_decimal': 1,
                u'msg': u'',
                u'ok': True},),
              ({u'expect': u'b',
                u'grade_decimal': 1,
                u'msg': u'',
                u'ok': True},)],
  u'grade_decimal': 1,
  u'msg': u'',
  u'ok': True},)
```
Hence, we gain the ability to apply grade_decimals and messages at higher levels, so something like
```python
grader = SingleListGrader(
    answers={'expect':['a', 'b'], 'msg':'Ok!},
    subgrader=StringGrader()
)
```
is accepted.

jolyonb commented:
This is now working! It was much easier, cleaner and quicker than I had anticipated!

I still need to write tests...

jolyonb commented:
@ChristopherChudzicki Ready for review. I think you'll like it. For a major overhaul, this is a whole lot simpler than you may expect! (pun intended ;-)

jolyonb commented on mitxgraders/listgrader.py:
```diff
@@ -758,9 +756,18 @@ def check_response(self, answer, student_input, **kwargs):
+                                           partial_credit=self.config['partial_credit'])
+
+        # If the result is ok, display the message for this answer
+        if result['ok'] is True and msg != '':
```
This is the only thing I'm in two minds about. I'd like the message to display if the student responses matched their expect values, and I'm using 'ok' as a proxy for this. However, if grade_decimal for a response is less than 1, then I won't catch it. There's just no way of reporting that the entire answer matched something, irregardless of the grade_decimal that was assigned to it... Any thoughts?

jolyonb replied:
Something similar to this. You want `nice job!` to display for `cat, dog` and `cat, DOG`, even though `dog` doesn't quite get full credit. You can't associate a message with `dog` or `DOG` directly, because then `unicorn, dog` and `unicorn, DOG` will both get that message.

ChristopherChudzicki replied:
In the  situation above, if I wanted to give a message on “cat, DOG” but not “unicorn, DOG” then I  would use a tuple of answers in the outermost SingleListGrader. 

Does that make sense? Sorry I can’t give a code example now, on my phone

jolyonb replied:
Here's what I have in mind exactly:
```python
grader = SingleListGrader(
    answers=(
        {
            'expect': [('X', {'expect': 'x', grade_decimal:0.5}),
                       ('Y', {'expect': 'y', grade_decimal:0.5})], 
            'msg': 'Good pairing!'
        },
        {
            'expect': [('W', {'expect': 'w', grade_decimal:0.5}),
                       ('Z', {'expect': 'z', grade_decimal:0.5})], 
            'msg': 'Good pairing!'
        }
    ),
    subgrader=StringGrader()
)
```
The message should appear for `X, Y`, `x, Y`, `X, y`, and `x, y`, or similarly for `W/w` and `Z/z`. If you get the pairing wrong though, say, `X, Z`, you don't get a message.

Currently, the message will only appear for `X, Y` or `W,Z`, because those are the only combination that awards full credit.

Happy to wait for you to take a look tonight - go and enjoy your day!

ChristopherChudzicki replied:
Ah, I understand. Well, to your original question:

> This is the only thing I'm in two minds about. I'd like the message to display if the student responses matched their expect values, and I'm using 'ok' as a proxy for this. However, if grade_decimal for a response is less than 1, then I won't catch it. There's just no way of reporting that the entire answer matched something, irregardless of the grade_decimal that was assigned to it... Any thoughts?

I agree that would be a more logical way to show messages.  I think that `input_list` (before consolidation) contains all the information you need. Something like

```python
if all(r['grade_decimal']>0 for r in input_list) and msg != '':
    result['msg'] = msg if result['msg'] == '' else result['msg'] + '\n' + msg
```
(or check that r['ok'] is truthy should be fine, too)

jolyonb replied:
I considered this, but I think it fails for nested `SingleListGrader`s, where partial credit might just mean that you got one wrong?

jolyonb commented:
@ChristopherChudzicki Just wondering if you plan to review this, or should I leave it sit a few days and then self-review?

ChristopherChudzicki commented:
I’ll try to review it tonight… I started reading through it this morning… But didn’t finish. If I don’t review it by tomorrow morning then I probably won’t for at least a week, so happy to have you do it.

But, I am reasonably confident I can do it tonight

ChristopherChudzicki commented on mitxgraders/listgrader.py:
```diff
@@ -633,37 +640,29 @@ def schema_config(self):
+        # validate_expect is shadowed, and we do post-processing in post_schema_validation
 
     # Make sure that the ItemGrader validate_expect isn't used
     validate_expect = None
```
This `validate_expect = None` should be deleted now, right?

jolyonb replied:
Absolutely. Didn't even realize that was there.

ChristopherChudzicki commented on mitxgraders/baseclasses.py:
```diff
@@ -495,6 +502,21 @@ def schema_answers(self, answer_tuple):
         schema = Schema((self.validate_single_answer,))
         return schema(answer_tuple)
 
+    def post_schema_validation(self, answer_tuple):
```
Is `post_schema_answer_validation` too long a name? Some graders do post-schema validation that is not answer validation, so it is "post schema validation" but it is not covered by this function.

jolyonb replied:
Indeed, this is only post schema answer validation (only required because of the possibility of inferring answers). I thought about your name, but it was already leading to overly long lines! Maybe I'll call it `post_schema_ans_val`, which isn't quite so long, but still conveys the relevant information.

jolyonb commented:
Ok, I fixed those two little things. I basically did your suggestion for checking whether or not to add the message, but I had to complicate it somewhat in order to account for nested SingleListGraders... Basically, I added an extra field to the return dictionary that recorded the required information, and then stripped it out in the __call__ method.

jolyonb commented:
@ChristopherChudzicki Updated and ready for final review.

ChristopherChudzicki commented on mitxgraders/listgrader.py:
```diff
@@ -761,9 +758,25 @@ def check_response(self, answer, student_input, **kwargs):
+        all_awarded = False
+        if not isinstance(self.config['subgrader'], SingleListGrader):
+            # Check to see if all items were awarded credit
+            if all(item['grade_decimal'] > 0 for item in input_list):
```
Minor comment: here and below could just do 

`all_awarded=all(item['grade_decimal'] > 0 for item in input_list)`

I believe no need for conditional. Also no need for initialization to False above.

jolyonb replied:
Good point - fixed.

ChristopherChudzicki commented:
Nice resolution to the message issue!
